### PR TITLE
Custom middlewares

### PR DESCRIPTION
**Pull request by kolinkovlad**
Added custom middlewares: AuthRequiredMiddleware, which realize redirect according to permissions and SwitchSessionDataMiddleware, which realize switching session data between stored and current sessions.

### CODE REVIEW

File: CultureAnalyzer/middleware.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+                Session.objects.update(
+                    session_key=request.session.session_key,
+                    session_data=encoded_data, expire_date=expire_date)
+
+            request.user.logged_in_user.session_key = request.session.session_key
+            request.user.logged_in_user.save()
+
+        response = self.get_response(request)
+
+        return response
```
Comment (rostIvan) [inline CultureAnalyzer/middleware.py]: You can to extract some of these lines in `__call__` and divide it on methods

File: CultureAnalyzer/middleware.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    def __call__(self, request):
+        return self.get_response(request)
+
+
+class SwitchSessionDataMiddleware:
+    def __init__(self, get_response):
+        self.get_response = get_response
+
+    def __call__(self, request):
+
```
Comment (MrKreg) [inline CultureAnalyzer/middleware.py]: Reformat your code with IDE!

File: CultureAnalyzer/middleware.py [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+            if stored_session_key and stored_session_key != request.session.session_key:
+                stored_session_data = Session.objects.get(
+                    session_key=stored_session_key).get_decoded()
+                expire_date = request.session.get_expiry_date()
+                encoded_data = SessionStore().encode(stored_session_data)
+                Session.objects.get(session_key=stored_session_key).delete()
+                Session.objects.update(
+                    session_key=request.session.session_key,
+                    session_data=encoded_data, expire_date=expire_date)
```
Comment (MrKreg) [inline CultureAnalyzer/middleware.py]: It would be better to wrap this piece of code into transaction.

**Review by MrKreg — CHANGES_REQUESTED** [review]

File: CultureAnalyzer/middleware.py [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+class SwitchSessionDataMiddleware:
+    def __init__(self, get_response):
+        self.get_response = get_response
+
+    def __call__(self, request):
+        self._check_session(request)
+
+        response = self.get_response(request)
+
+        return response
```
Comment (rostIvan) [inline CultureAnalyzer/middleware.py]: simplify
``` python 
return self.get_response(request) 
```

File: CultureAnalyzer/middleware.py [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+        """
+        if request.user.is_authenticated:
+            current_session_key = request.session.session_key
+            stored_session_key = request.user.logged_in_user.session_key
+
+            if stored_session_key and stored_session_key != current_session_key:
+                self._switch_session_data(request, stored_session_key)
+            request.user.logged_in_user.session_key = current_session_key
+            request.user.logged_in_user.save()
+
```
Comment (rostIvan) [inline CultureAnalyzer/middleware.py]: I'm not sure, but maybe the next lines should be also inside `transaction`
https://github.com/IF-093-Python/CultureAnalyzer/blob/df68f76ca428c68e7e6e28f0648aaf939786920e/CultureAnalyzer/middleware.py#L46-L47

You can to append this part of code to `_switch_session_data` method

File: CultureAnalyzer/middleware.py (hunk 4 above)
Comment (kolinkovlad) [reply CultureAnalyzer/middleware.py]: something like: return = self.get_response(request) ?

**Review by MrKreg — APPROVED** [review]

**Review by falled10 — APPROVED** [review]

**Pull request merged**